crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs

```diff
@@ -0,0 +1,258 @@
+use itertools::Itertools;
+
+use ruff_diagnostics::{Applicability, Diagnostic, Edit, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::{visitor::Visitor, Expr, ExprSubscript};
+use ruff_python_ast::{StmtClassDef, StmtFunctionDef};
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+use crate::settings::types::PythonVersion;
+
+use super::{DisplayTypeVars, TypeVar, TypeVarReferenceVisitor};
+
+/// ## What it does
+///
+/// Checks for use of standalone type variables and parameter specifications in generic functions
+/// and classes.
+///
+/// ## Why is this bad?
+///
+/// Special type parameter syntax was introduced in Python 3.12 by [PEP 695] for defining generic
+/// functions and classes. This syntax is easier to read and provides cleaner support for generics.
+///
+/// ## Known problems
+///
+/// [PEP 695] uses inferred variance for type parameters, instead of the `covariant` and
+/// `contravariant` keywords used by `TypeVar` variables. As such, rewriting a `TypeVar` variable to
+/// an in-line type parameter may change its variance.
+///
+/// The rule currently excludes cases where it conceptually should be able to give a diagnostic. In
+/// particular, it skips generic classes with multiple base classes, and it skips generic methods in
+/// generic or non-generic classes.
+///
+/// ## Example
+///
+/// ```python
+/// from typing import TypeVar
+///
+/// T = TypeVar("T")
+///
+///
+/// class GenericClass(Generic[T]): ...
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// class GenericClass[T]: ...
+/// ```
+///
+/// [PEP 695]: https://peps.python.org/pep-0695/
+#[derive(ViolationMetadata)]
+pub(crate) struct NonPEP695TypeParameter {
+    name: String,
+    generic_kind: GenericKind,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum GenericKind {
+    GenericClass,
+    GenericFunction,
+}
+
+impl Violation for NonPEP695TypeParameter {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Always;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let NonPEP695TypeParameter { name, generic_kind } = self;
+        match generic_kind {
+            GenericKind::GenericClass => {
+                format!("Generic class `{name}` uses `Generic` subclass instead of type parameters")
+            }
+            GenericKind::GenericFunction => {
+                format!("Generic function `{name}` should use type parameters")
+            }
+        }
```

Comment:
Another reason why it might be good for these to be implemented as separate rules: the rules table in the docs [here](https://docs.astral.sh/ruff/rules/) is generated by this method; it will use the contents of the first `format!()` call in this method for this rule's "message" column. Only displaying the first `format!()` call and not the second would give a slightly misleading impression of what the rule does, in this case!